src/java.base/share/classes/jdk/internal/reflect/NativeMethodAccessorImpl.java

```diff
@@ -49,16 +62,27 @@ public Object invoke(Object obj, Object[] args)
         // found from the generated bytecode.
         if (++numInvocations > ReflectionFactory.inflationThreshold()
                 && !method.getDeclaringClass().isHidden()
-                && !ReflectUtil.isVMAnonymousClass(method.getDeclaringClass())) {
-            MethodAccessorImpl acc = (MethodAccessorImpl)
-                new MethodAccessorGenerator().
-                    generateMethod(method.getDeclaringClass(),
-                                   method.getName(),
-                                   method.getParameterTypes(),
-                                   method.getReturnType(),
-                                   method.getExceptionTypes(),
-                                   method.getModifiers());
-            parent.setDelegate(acc);
+                && !ReflectUtil.isVMAnonymousClass(method.getDeclaringClass())
+                && ACCESSOR_GENERATED.compareAndSet(this, false, true)) {
```

Comment:
As the micro-optimization, checking that `accessor_generated` is `false` before attempting a (potentially contended) CAS might fit better. (See https://en.wikipedia.org/wiki/Test_and_test-and-set).

Reply:
Agree this could benefit performance. Grep jdk codes (grep -R  compareAndSet *), not found codes with test and test-and-set pattern.

compareAndSet is intrinsic and maybe it's better let runtime/compiler generate Test_and_test-and-set code sequence to solve all compareAndSet operations.